File: volume_1/chapter_4/stream.js
/** Streaming helpers for chapter web APIs. */

function extractPythonStyleContent(payload) {
    const marker = 'content=';
    const start = payload.indexOf(marker);
    if (start < 0) return null;

    const quoteIndex = start + marker.length;
    const quote = payload[quoteIndex];
    if (quote !== '"' && quote !== "'") return null;

    let i = quoteIndex + 1;
    let escaped = false;
    let value = '';

    while (i < payload.length) {
        const char = payload[i];
        if (escaped) {
            switch (char) {
                case 'n':
                    value += '\n';
                    break;
                case 'r':
                    value += '\r';
                    break;
                case 't':
                    value += '\t';
                    break;
                default:
                    value += char;
                    break;
            }
            escaped = false;
            i += 1;
            continue;
        }

        if (char === '\\') {
            escaped = true;
            i += 1;
            continue;
        }

        if (char === quote) {
            const remaining = payload.slice(i + 1);
            if (remaining.includes('additional_kwargs=')) {
                return value;
            }
            return null;
        }

        value += char;
        i += 1;
    }

    return null;
}

export function normalizeResponseText(payload) {
    if (payload == null) return '';

    if (typeof payload === 'string') {
        const extracted = extractPythonStyleContent(payload);
        if (typeof extracted === 'string') {
            return extracted;
        }

        try {
            const maybeJson = JSON.parse(payload);
            if (maybeJson && typeof maybeJson === 'object') {
                for (const key of ['answer', 'distilled', 'content', 'text', 'message', 'summary']) {
                    const value = maybeJson[key];
                    if (typeof value === 'string' && value.trim()) return value;
                }
            }
        } catch {
            // not JSON text, keep original payload
        }

        return payload;
    }

    if (typeof payload === 'object') {
        for (const key of ['content', 'text', 'message', 'answer', 'final_answer', 'distilled', 'summary']) {
            const value = payload[key];
            if (typeof value === 'string' && value.trim()) return value;
        }
        return JSON.stringify(payload);
    }

    return String(payload);
}

export function chunkText(text, chunkSize = 28) {
    const clean = text || '';
    if (!clean) return [''];

    const chunks = [];
    for (let index = 0; index < clean.length; index += chunkSize) {
        chunks.push(clean.slice(index, index + chunkSize));
    }
    return chunks;
}

File: volume_1/chapter_4/stream.py
"""Streaming helpers for chapter web APIs."""

from __future__ import annotations

import asyncio
import ast
import re
import json
from typing import AsyncIterator, Iterable


def normalize_response_text(payload) -> str:
    """Normalize model responses to plain text for streaming transports."""
    if payload is None:
        return ""
    if isinstance(payload, str):
        content_match = re.search(
            r"content=(['\"])((?:\\.|(?!\1).)*)\1\s+additional_kwargs=",
            payload,
            flags=re.DOTALL,
        )
        if content_match:
            raw_quoted_content = f"{content_match.group(1)}{content_match.group(2)}{content_match.group(1)}"
            try:
                decoded = ast.literal_eval(raw_quoted_content)
                if isinstance(decoded, str):
                    return decoded
            except Exception:
                return content_match.group(2)

        try:
            maybe_json = json.loads(payload)
            if isinstance(maybe_json, dict):
                for key in ("answer", "distilled", "content", "text", "message", "summary"):
                    value = maybe_json.get(key)
                    if isinstance(value, str) and value.strip():
                        return value
        except Exception:
            pass

        return payload
    if isinstance(payload, dict):
        for key in ("content", "text", "message", "answer", "final_answer", "distilled", "summary"):
            value = payload.get(key)
            if isinstance(value, str) and value.strip():
                return value
    if hasattr(payload, "content") and isinstance(payload.content, str):
        return payload.content
    return str(payload)


def chunk_text(text: str, chunk_size: int = 28) -> Iterable[str]:
    """Chunk text so non-streaming providers can still emit incremental tokens."""
    clean = text or ""
    if not clean:
        yield ""
        return

    for i in range(0, len(clean), chunk_size):
        yield clean[i : i + chunk_size]


async def iter_text_chunks(text: str, chunk_size: int = 28, delay_seconds: float = 0.0) -> AsyncIterator[str]:
    """Yield text chunks asynchronously with optional pacing delay."""
    for part in chunk_text(text, chunk_size=chunk_size):
        if delay_seconds > 0:
            await asyncio.sleep(delay_seconds)
        yield part
